### ISSUE TITLE
Move RESP serialisation into RedisExecutor service

### ISSUE THREAD

**Issue by quelgar** [zio/zio-redis#136]
I think some changes to `Interpreter` are needed.

The problem is that the current method of receiving responses isn't reliable, and it can't support pipelining. The `unsafeReceive` method just keeps reading until it only partially fills its buffer, assuming that means it has now read the entire response. There's a few ways this can go wrong, but the easiest way to make it fail when using localhost is to `SET` a key with a 1015 byte value. If you `GET` that with zio-redis, the read will very likely hang because the response happens to fit exactly in the buffer.

I can't see a way to fix this with the current `RedisExecutor` API, which accepts and returns RESP serialized strings. The problem with this is you can't reliably tell where a response ends without actually deserializing.

I propose to fix this by moving RESP serialization/deserialization into the executor. This is generic serialization that doesn't depend on which command is being sent. The RESP serialization can be represented with something like this:

```
sealed trait RespValue

object RespValue {
  final case class SimpleString(value: String) extends RespValue

  final case class Error(value: String) extends RespValue

  final case class Integer(value: Long) extends RespValue

  final case class BulkString(value: Chunk[Byte]) extends RespValue

  final case class Array(values: List[RespValue]) extends  RespValue
  
  case object NullString extends RespValue
}
```

and the executor service could be:

```
trait Service {

      def execute(command: RespValue.Array): IO[RedisError, RespValue]

    }
```

The receive then keeps reading bytes until the deserialization produces a complete response value, which can then be provided to the promise, and then moves on to reading the next response (I'm thinking the deserializaition can be implemented as `ZTransducer`). This will also allow support for pipelining, where the responses will not remotely align with the read buffer.

This will require changes to the `Input` serialization and `Output` deserialization as it exists today, but it should make them a little simpler and separate concerns nicely. For example, `DurationSecondsOutput` can just deal with converting the integer value to a duration, and will no longer need to know the RESP details like the string starting with `:` and parsing the rest to an int.

Thoughts?

**Comment by mijicd** [comment]
This proposal makes sense, but it will require significant changes in the current codebase. It makes sense to prioritize it over transactions and streams to avoid duplicated work. Also, it would be great if we can somehow parallelize it.

**Comment by mijicd** [comment]
Closed by #139
